azurerm/resource_arm_container_group.go

```diff
@@ -368,3 +490,46 @@ func expandContainerEnvironmentVariables(input interface{}) *[]containerinstance
 
 	return &output
 }
+
+func expandContainerVolumes(input interface{}, containerGroupVolumes *[]containerinstance.Volume) *[]containerinstance.VolumeMount {
+	volumesRaw := input.([]interface{})
+
+	if len(volumesRaw) == 0 {
+		return nil
+	}
+
+	volumeMounts := make([]containerinstance.VolumeMount, 0, len(volumesRaw))
+
+	for _, volumeRaw := range volumesRaw {
+		volumeConfig := volumeRaw.(map[string]interface{})
+
+		name := volumeConfig["name"].(string)
+		mountPath := volumeConfig["mount_path"].(string)
+		readOnly := volumeConfig["read_only"].(bool)
+		shareName := volumeConfig["share_name"].(string)
+		storageAccountName := volumeConfig["storage_account_name"].(string)
+		storageAccountKey := volumeConfig["storage_account_key"].(string)
+
+		vm := containerinstance.VolumeMount{
+			Name:      &name,
+			MountPath: &mountPath,
+			ReadOnly:  &readOnly,
+		}
+
+		volumeMounts = append(volumeMounts, vm)
+
+		cv := containerinstance.Volume{
+			Name: &name,
+			AzureFile: &containerinstance.AzureFileVolume{
+				ShareName:          &shareName,
+				ReadOnly:           &readOnly,
+				StorageAccountName: &storageAccountName,
+				StorageAccountKey:  &storageAccountKey,
+			},
+		}
+
+		*containerGroupVolumes = append(*containerGroupVolumes, cv)
```

Comment:
I'd argue this would be clearer if we instead returned multiple objects from this method, rather than appending to the input argument i.e.

```
func foo(bar string) (*[]containerinstance.Volume, *[]containerinstance.VolumeMount) {
  return volumes, volumeMounts
}
```

Reply:
It has to be this way since the containerGroupVolumes is a collection of all volumes across all containers.

Reply:
Reading the code above, it takes understanding of the inner working of the method to know that both a value passed in will be modified, and a return value will come out - in general methods tend to do one or the other (and I'd argue the latter is easier to understand). We should be able to achieve the same thing by returning the value and then setting it - for instance:

```
func foo() {
  containerGroupVolumes := make([]containerinstance.Volume, 0)
  
  for i := 0; i < 9; i++ {
    mounts, containerGroupVolumes := expandContainerVolumes(interface{}, &containerGroupVolumes)
    // mounts is local
    // containerGroupVolumes is appended too
  }
}

func expandContainerVolumes(input interface{}, containerGroupVolumes *[]containerinstance.Volume) (*[]containerinstance.VolumeMount, containerGroupVolumes *[]containerinstance.Volume) {
  mounts := make([]containerinstance.VolumeMount, 0)
  volumes := *containerGroupVolumes
  volumes = append(volumes, ...)
  return &mounts, &volumes
}
```

Reply:
Done.